### PR TITLE
[SPARK-28456][SQL] Add a public API `Encoder.makeCopy` to allow creating Encoder without touching Scala Reflection

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Because `Encoder` is not thread safe, the user cannot reuse an `Encoder` in multiple `Dataset`s. However, creating an `Encoder` for a complicated class is slow due to Scala Reflection. To eliminate the cost of Scala Reflection, right now I usually use the private API `ExpressionEncoder.copy` as follows:

```scala
object FooEncoder {
  private lazy val _encoder: ExpressionEncoder[Foo] = ExpressionEncoder[Foo]()
  implicit def encoder: ExpressionEncoder[Foo] = _encoder.copy()
}
```

This PR proposes a new method `makeCopy` in `Encoder` so that the above codes can be rewritten using public APIs.

```scala
object FooEncoder {
  private lazy val _encoder: Encoder[Foo] = Encoders.product[Foo]()
  implicit def encoder: Encoder[Foo] = _encoder.makeCopy
}
```

The method name is consistent with `TreeNode.makeCopy`.

## How was this patch tested?

Jenkins

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @marmbrus @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala [hunk 1]
```diff
@@ -452,6 +452,10 @@ class ExpressionEncoderSuite extends CodegenInterpretedPlanTest with AnalysisTes
     }
   }
 
+  encodeDecodeTest("foo" -> 1L, "copied encoder") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala]: @zsxwing, can we move this test up over `testOverflowingBigNumeric`? I think private helper method look intentionally grouped below here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala [hunk 2]
```diff
@@ -80,4 +80,10 @@ trait Encoder[T] extends Serializable {
    * A ClassTag that can be used to construct an Array to contain a collection of `T`.
    */
   def clsTag: ClassTag[T]
+
+  /**
+   * Create a copied [[Encoder]]. The implementation may just copy internal reusable fields to speed
+   * up the [[Encoder]] creation.
+   */
+  def copyEncoder: Encoder[T]
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: How about `makeCopy`? Looks like `TreeNode.makeCopy` tried to avoid `copy`.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: Seems good `makeCopy`.

Comment (zsxwing) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: @HyukjinKwon `makeCopy` sounds better to me as well. Updated.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**